src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -357,6 +358,159 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             return expandedBytes;
         }
 
+        private static unsafe string UrlEncode(char* s, int length)
+        {
+            int numUnsafe = 0;
+            int numSpaces = 0;
+            for (int i = 0; i < length; i++)
+            {
+                char c = s[i];
+                if (c == ' ')
+                {
+                    numSpaces++;
+                }
+                else if (!IsUrlSafeChar(s[i]))
+                {
+                    numUnsafe += GetByteCount(s, i, length);
```

Comment:
That's a fair shout. I just wish the code for UTF8Encoding was in corefx not coreclr so I could contribute